subsystems/TypeHandlerLibrary/src/main/java/org/terasology/persistence/typeHandling/bytebuffer/ByteBufferPersistedData.java

```diff
@@ -0,0 +1,236 @@
+        this(byteBuffer, byteBuffer.position(), type);
+    }
+
+
+    public ByteBuffer getData() {
+        return byteBuffer;
+    }
+
+    @Override
+    public String getAsString() {
+        if (!isString()) {
+            throw new ClassCastException("it is not string");
+        }
+        resetPosition();
+        int size = byteBuffer.getInt();
+        byte[] bytes = new byte[size];
+        byteBuffer.get(bytes);
+        return new String(bytes, StandardCharsets.UTF_8);
+    }
+
+    private void resetPosition() {
+        byteBuffer.position(position + (typeForced ? 0 : 1));
+    }
+
+    @Override
+    public double getAsDouble() {
+        if (!isNumber()) {
+            throw new ClassCastException("It is not number");
+        }
+        resetPosition();
+        switch (type) {
+            case INTEGER:
+                return getData().getInt();
+            case FLOAT:
+                return getData().getFloat();
+            case DOUBLE:
+                return getData().getDouble();
+            case LONG:
+                return (double) getData().getLong();
+        }
+        throw new ClassCastException("Data is not a number");
+    }
+
+    @Override
+    public float getAsFloat() {
+        if (!isNumber()) {
+            throw new ClassCastException("It is not number");
+        }
+        resetPosition();
+        switch (type) {
+            case INTEGER:
+                return (float) getData().getInt();
+            case FLOAT:
+                return getData().getFloat();
+            case DOUBLE:
+                return (float) getData().getDouble();
+            case LONG:
+                return (float) getData().getLong();
+        }
+        throw new ClassCastException("Data is not a number");
+    }
+
+    @Override
+    public int getAsInteger() {
+        if (!isNumber()) {
+            throw new ClassCastException("It is not number");
+        }
+        resetPosition();
+        switch (type) {
+            case INTEGER:
+                return getData().getInt();
+            case FLOAT:
+                return (int) getData().getFloat();
+            case DOUBLE:
+                return (int) getData().getDouble();
+            case LONG:
+                return (int) getData().getLong();
+        }
+        throw new ClassCastException("Data is not a number");
+    }
+
+    @Override
+    public long getAsLong() {
+        if (!isNumber()) {
+            throw new ClassCastException("It is not number");
+        }
+        resetPosition();
+        switch (type) {
+            case INTEGER:
+                return getData().getInt();
+            case FLOAT:
+                return (long) getData().getFloat();
+            case DOUBLE:
+                return (long) getData().getDouble();
+            case LONG:
+                return getData().getLong();
+        }
+        throw new ClassCastException("Data is not a number");
```

Comment:
```suggestion
        throw new ClassCastException(String.format("source is is not of type number: %s", type.name());
```

maybe just have classCastException also take a type.